casper/src/main/scala/coop/rchain/casper/Validate.scala

```diff
@@ -418,6 +422,43 @@ object Validate {
       } yield Left(InvalidDeployCount)
     }
 
+  // TODO: Double check logic with validator rotation
+  def lastSequenceNumber[F[_]: Monad: Log: BlockStore](
+      b: BlockMessage,
+      genesis: BlockMessage
+  ): F[Either[InvalidBlock, ValidBlock]] =
+    for {
+      maybeMainParent <- ProtoUtil.mainParent[F](b)
+      result <- maybeMainParent match {
+                 case Some(mainParent) =>
+                   val mainParentLastSequenceNumberMap =
+                     ProtoUtil.toLastSequenceNumberMap(mainParent.lastSequenceNumber)
+                   val computedLastSequenceNumber =
+                     mainParentLastSequenceNumberMap.updated(b.sender, b.seqNum)
+                   if (b.lastSequenceNumber == ProtoUtil.toLastSequenceNumber(
+                         computedLastSequenceNumber
+                       )) {
+                     Applicative[F].pure(Right(Valid))
+                   } else {
+                     val logMessage =
+                       s"computed last sequence number is not equal to given last sequence number."
+                     for {
+                       _ <- Log[F].warn(ignore(b, logMessage))
+                     } yield Left(InvalidLastSequenceNumber)
+                   }
+                 case None =>
+                   if (b == genesis) {
+                     Applicative[F].pure(Right(Valid))
+                   } else {
+                     val logMessage =
+                       s"main parent is missing even though block is not genesis."
+                     for {
+                       _ <- Log[F].warn(ignore(b, logMessage))
+                     } yield Left(InvalidLastSequenceNumber)
+                   }
+               }
+    } yield result
```

Comment:
I find this function difficult to read. can you make it more readable?

Reply:
Hmm should I extract the two cases?

Reply:
maybe:
```
  // TODO: Double check logic with validator rotation
  def lastSequenceNumber[F[_]: Monad: Log: BlockStore](
    b: BlockMessage,
    genesis: BlockMessage
  ): F[Either[InvalidBlock, ValidBlock]] = {
    def valid = (Valid:ValidBlock).asRight[InvalidBlock].pure[F]
    def invalid(logMessage: String) = Log[F].warn(ignore(b, logMessage)).as((InvalidLastSequenceNumber: InvalidBlock).asLeft[ValidBlock])
    def calculateLastSequenceNumber(mainParent: BlockMessage) = {
      val mainParentLastSequenceNumberMap =
        ProtoUtil.toLastSequenceNumberMap(mainParent.lastSequenceNumber)
      val computedLastSequenceNumber =
        mainParentLastSequenceNumberMap.updated(b.sender, b.seqNum)
      ProtoUtil.toLastSequenceNumber(computedLastSequenceNumber)
    }

    if (b == genesis)
      valid
    else
      ProtoUtil.mainParent[F](b) >>= {
        case Some(mainParent) =>
          if (b.lastSequenceNumber == calculateLastSequenceNumber(mainParent))
            valid
          else
            invalid("computed last sequence number is not equal to given last sequence number.")

        case None =>
          invalid("main parent is missing even though block is not genesis.")
      }
  }
```

Reply:
I'm thinking about something like this:
```scala
  def lastSequenceNumber[F[_]: Monad: Log: BlockStore](
      b: BlockMessage,
      genesis: BlockMessage
  ): F[Either[InvalidBlock, ValidBlock]] = {

    def computeLastSequenceNumber(mainParent : BlockMessage) = {
      val mainParentLastSequenceNumberMap =
        ProtoUtil.toLastSequenceNumberMap(mainParent.lastSequenceNumber)
      val computedLastSequenceNumberMap =
        mainParentLastSequenceNumberMap.updated(b.sender, b.seqNum)
      ProtoUtil.toLastSequenceNumber(computedLastSequenceNumberMap)
    }

    def computeErrors(maybeMainParentSequenceNumber : Option[Seq[LastSequenceNumber]]) =
      maybeMainParentSequenceNumber match {
         case Some(computedLastSequenceNumber) if (b.lastSequenceNumber != computedLastSequenceNumber) =>
           Some("computed last sequence number is not equal to given last sequence number.")
         case None if (b != genesis) => 
           Some("main parent is missing even though block is not genesis.")
         case _ => None
      }

    for {
      maybeParent <- ProtoUtil.mainParent[F](b)
    
      result <- computeErrors(maybeParent.map(computeLastSequenceNumber)) match {
        case None => Right(Valid).pure[F]
          
        case Some(errorMsg) => 
          Log[F].warn(ignore(b, errorMsg)) *> 
            Left(InvalidLastSequenceNumber).pure[F]
      }
    } yield result
  }```

Reply:
...or the for could be written like this:
```scala
    for {
      maybeParent <- ProtoUtil.mainParent[F](b)
      maybeLastSequenceNumber = maybeParent.map(computeLastSequenceNumber)
      result <-
        computeErrors(maybeLastSequenceNumber)
          .traverse(errorMsg => Log[F].warn(ignore(b, errorMsg)) *> Left(InvalidLastSequenceNumber).pure[F])
    } yield result.getOrElse(Right(Valid))
```